extensions/amp-ad/0.1/amp-ad-xorigin-iframe-handler.js

```diff
@@ -161,39 +184,55 @@ export class AmpAdXOriginIframeHandler {
       this.element_.appendChild(this.iframe);
       return Promise.resolve();
     }
-    // Set iframe initially hidden which will be removed on load event +
-    // post message.
+
+    // Set iframe initially hidden which will be removed on render-start or
+    // load, whichever is earlier.
     setStyle(this.iframe, 'visibility', 'hidden');
     this.element_.appendChild(this.iframe);
+    const visibilityPromise = Promise.race([
+      renderStartPromise,
+      iframeLoadPromise,
+      timer.promise(VISIBILITY_TIMEOUT),
+    ]).then(() => {
+      if (this.iframe) {
+        setStyle(this.iframe, 'visibility', '');
+        if (this.baseInstance_.emitLifecycleEvent) {
+          this.baseInstance_.emitLifecycleEvent('adSlotUnhidden');
+        }
+      }
+    });
 
-    return timerFor(this.baseInstance_.win).timeoutPromise(TIMEOUT_VALUE,
-        this.adResponsePromise_,
-        'timeout waiting for ad response').catch(e => {
-          this.noContent_();
-          user().warn('AMP-AD', e);
-        }).then(() => {
-          if (this.iframe) {
-            setStyle(this.iframe, 'visibility', '');
-            if (this.baseInstance_.emitLifecycleEvent) {
-              this.baseInstance_.emitLifecycleEvent('adSlotUnhidden');
-            }
-          }
-        });
+    // The actual ad load is eariliest of iframe.onload event and no-content.
+    const adLoadPromise = Promise.race([
+      iframeLoadPromise,
+      noContentPromise,
+    ]);
+    return adLoadPromise.then(() => {
+      // If ad loading has succeeded, so should either visibilityPromise or
+      // no-content. Return it here to ensure that rendering has been fully
+      // processed when layout promise is complete.
+      return Promise.race([visibilityPromise, noContentPromise]);
```

Comment:
At this point, either `iframeLoadPromise` or `noContentPromise` is resolved.
In case of `noContentPromise` is resolved, this `Promise.race` is immediately resolved.
In case of `iframeLoadPromise` is resolved, `visibilityPromise` is resolved as well, because it's race of `iframeLoadPromise` with others. So this `Promise.race` is also immediately resolved.

So, this Promise.race will always be immediately resolved, meaning we don't need it.